keras/engine/network.py

```diff
@@ -1028,8 +1039,16 @@ def process_layer(layer_data):
             for layer_data in config['layers']:
                 layer = created_layers[layer_data['name']]
                 if layer in unprocessed_nodes:
-                    for node_data in unprocessed_nodes.pop(layer):
-                        process_node(layer, node_data)
+                    # Process nodes in order, if an error save the
+                    # remaining nodes to process later.
+                    node_data_list = unprocessed_nodes.pop(layer)
+                    while node_data_list:
+                        node_data = node_data_list.pop(0)
```

Comment:
Here you remove an element from `node_data_list`, process it, and if something is wrong, put it back in the list. 

Could you instead process the first element, and if there is no error, remove it from the list? Same for the `unpreprocessed_nodes`, don't pop the layer and put it back if there is an error. Remove the layer only when everything is processed. It will be much easier to follow the progression with a debugger and there will be no temporary list (I'm talking about the variable `node_data_list`). 

The logic of your algorithm will be followed through the variable `unpreprocessed_nodes` gradually loosing elements.

Reply:
I might add that it will be easier to notice infinite loops by doing this.

Reply:
I have reworked the logic so elements are not removed while processing. Now the list is processed in order, if there is an error the loop will terminate and the unprocessed items are stored back in the `unprocessed_nodes` object. Otherwise, the `layer` will be removed from the `unprocessed_nodes` object.

Hopefully this is close to what you were after 😄